apps/postgres-new/app/api/databases/[id]/upload/route.ts

```diff
@@ -1,30 +1,83 @@
 import { S3Client } from '@aws-sdk/client-s3'
 import { Upload } from '@aws-sdk/lib-storage'
-import { NextRequest } from 'next/server'
+import { NextRequest, NextResponse } from 'next/server'
 import { createGzip } from 'zlib'
 import { Readable } from 'stream'
+import { createClient } from '~/utils/supabase/server'
+import { createScramSha256Data } from 'pg-gateway'
+import { generateDatabasePassword } from '~/utils/generate-database-password'
 
-const wildcardDomain = process.env.WILDCARD_DOMAIN ?? 'db.example.com'
+const wildcardDomain = process.env.NEXT_PUBLIC_WILDCARD_DOMAIN ?? 'db.example.com'
 const s3Client = new S3Client({ endpoint: process.env.S3_ENDPOINT, forcePathStyle: true })
 
-export async function POST(req: NextRequest, { params }: { params: { id: string } }) {
-  if (!req.body) {
-    return new Response(
-      JSON.stringify({
+export type DatabaseUploadResponse =
+  | {
+      success: true
+      data: {
+        username: string
+        password?: string
+        host: string
+        port: number
+        databaseName: string
+      }
+    }
+  | {
+      success: false
+      error: string
+    }
+
+export async function POST(
+  req: NextRequest,
+  { params }: { params: { id: string } }
+): Promise<NextResponse<DatabaseUploadResponse>> {
+  const supabase = createClient()
+
+  const {
+    data: { user },
+  } = await supabase.auth.getUser()
+
+  if (!user) {
+    return NextResponse.json(
+      {
         success: false,
-        error: 'Missing request body',
-      }),
+        error: 'Unauthorized',
+      },
+      { status: 401 }
+    )
+  }
+
+  const data = await req.formData()
+
+  const dump = data.get('dump') as File | null
+  const name = data.get('name') as string | null
+  const createdAt = data.get('created_at') as string | null
+
+  if (!dump || !name || !createdAt) {
+    return NextResponse.json(
       {
-        status: 400,
-      }
+        success: false,
+        error: 'Missing fields',
+      },
+      { status: 400 }
+    )
+  }
+
+  const dumpSizeInMB = dump.size / (1024 * 1024)
+  if (dumpSizeInMB > 100) {
+    return NextResponse.json(
+      {
+        success: false,
+        error: "You can't deploy a database that is bigger than 100MB",
+      },
+      { status: 413 }
     )
   }
 
   const databaseId = params.id
   const key = `dbs/${databaseId}.tar.gz`
 
   const gzip = createGzip()
```

Comment:
I noticed that sometimes PGlite produced a non-gzipped tar file (not sure what caused this). Should we try to detect this somehow or check mime type?

Reply:
In my case it's the opposite, I never had PGlite produce a gzipped tarball, that's why I gzip it in the route here.
Should we ping Sam about it?